activation/handler_v2.go

```diff
@@ -487,10 +489,50 @@ func (h *HandlerV2) equivocationSet(atx *wire.ActivationTxV2) ([]types.NodeID, e
 }
 
 type atxParts struct {
-	leaves         uint64
+	ticks          uint64
+	weight         uint64
 	effectiveUnits uint32
 }
 
+type nipostSize struct {
+	units uint32
+	ticks uint64
+}
+
+func (n *nipostSize) addUnits(units uint32) error {
+	sum, carry := bits.Add32(n.units, units, 0)
+	if carry != 0 {
+		return errors.New("units overflow")
+	}
+	n.units = sum
+	return nil
+}
+
+type nipostSizes []*nipostSize
+
+func (n nipostSizes) minTicks() uint64 {
+	return slices.MinFunc(n, func(a, b *nipostSize) int { return cmp.Compare(a.ticks, b.ticks) }).ticks
+}
+
+func (n nipostSizes) sumUp() (units uint32, weight uint64, err error) {
+	var totalEffectiveNumUnits uint32
+	var totalWeight uint64
+	for _, ns := range n {
+		sum, carry := bits.Add32(totalEffectiveNumUnits, ns.units, 0)
+		if carry != 0 {
+			return 0, 0, fmt.Errorf("total units overflow (%d + %d)", totalEffectiveNumUnits, ns.units)
+		}
+		totalEffectiveNumUnits = sum
+
+		hi, weight := bits.Mul64(uint64(ns.units), ns.ticks)
+		if hi != 0 {
+			return 0, 0, fmt.Errorf("weight overflow (%d * %d)", ns.units, ns.ticks)
+		}
+		totalWeight += weight
+	}
+	return totalEffectiveNumUnits, totalWeight, nil
+}
```

Comment:
I would then need to verify if `poetLeaves / tickSize` is within an `uint32`. I'd prefer to avoid making assumptions about "realistic" numbers as they might change with time or between networks. I'd rather work on the types and make assertions for the types' boundaries. These checks are not heavy and optimizing them would be premature at this point. If they show up in profiles in the future, we can think of optimizations.